### ISSUE TITLE
Paint Tool

### ISSUE THREAD

**Issue by ETdoFresh** [ETdoFresh/TileMapCreator#1]
Click a grid square... make the selected tile appear!

**Comment by onsclom** [comment]
I've been doing some planning since I know that there will be a lot of tools that need undo/redo in the future. My thinking is to make a list called `actions` which holds `GridAction` objects.

A GridAction object may look something like :
```python3
{
   # what tiles were changed to in this action
   new_tiles: [], # a list of GridTile objects
   # what tiles were previously before the action
   old_tiles: [] # a list of GridTile objects
}
```
Things like paint and erase would only have 1 new and 1 old GridTile object, but things like the line or bucket tool may have multiple of each.

A GridTile object may look something like:
```python3
{
    position: Vector2(x,y),
    tile_texture: ??? # could be null if representing an empty tile
}
```

The `actions` list acts as a history keeper which should help a lot with undo/redo features. Undo and redo for add/remove layers may be a little more complex since that involves keeping track of everything that was on that layer and the layer's relative position, but maybe that can be represented as its own object and stored in the `actions` list as well.

What do you think about this plan? Would this be the optimal way to go about this?

Alternatively, how memory intensive would it be to save the whole grid scene upon every action as shown from the [PackedScene doc page](https://docs.godotengine.org/en/stable/classes/class_packedscene.html)?
```python3
var scene = PackedScene.new()
# Only `node` and `rigid` are now packed.
var result = scene.pack(node)
if result == OK:
    var error = ResourceSaver.save("res://path/name.scn", scene)  # Or "user://..."
    if error != OK:
        push_error("An error occurred while saving the scene to disk.")
```
Maybe that would be an alternative? Not sure if that method would save all the data we needed however since it doesn't save script variables.

**Comment by ETdoFresh** [comment]
Yeah, I'd like to keep actions as abstract as possible... Let's say every action has at least 3 functions:
- undo()
- redo()
- perform(object)

The **object** parameter in _perform_() will be a dynamic value [most likely dictionary] that we pass.
for **paint grid** the object could be _{"Tile": tile, "Texture": texture}_
for **line** the object could be _["Tiles": tiles, "Texture": texture}_

> My thinking is to make a list called actions which holds GridAction objects.

Yeah, I like the idea of storing the old_value and new_value in the action. Here is a BaseAction and a RandomNumberGenerateAction I quickly made:

**base_action.gd**
``` gdscript
class_name BaseAction

func undo():
    pass

func redo():
    pass

func perform(_object):
    pass
```

**random_number_action.gd**
``` gdscript
class_name RandomNumberAction
extends BaseAction

var object
var old_value
var new_value

func undo():
    object.text = old_value

func redo():
    object.text = new_value

func perform(action_object):
    object = action_object
    old_value = object.text
    new_value = str(randi())
    object.text = new_value

```
_The perform(object) above is simply a Label Node_

> The actions list acts as a history keeper which should help a lot with undo/redo features.

Yes! An action list! Agreed! Another quick implementation:

**actions.gd**
``` gdscript
extends Node

var history = []
var index = -1

func undo():
    if index >= 0:
        history[index].undo()
        index -= 1

func redo():
    if history.size() == 0:
        return
    
    if index + 1 < history.size():
        history[index + 1].redo()
        index += 1

func perform(action_type, object):
    _clear_history_after(index)
    var action = action_type.new()
    history.append(action)
    action.perform(object)
    index += 1

func _clear_history_after(i):
    if i < history.size() - 1:
        for j in range(history.size() - 1, i, -1):
            history.remove(j)
```

> Alternatively, how memory intensive would it be to save the whole grid scene upon every action

Let's pass on this for now. I don't want to write to res:// or user:// (HDD) for all undo/redo actions.... although.... this will be a task later on to log this stuff.... soo, maybe.... but not yet.

**Comment by onsclom** [comment]
I really like those classes and plans you laid out Thanks for this really great write up. I will get start on this now.
